nomad/structs/network.go

```diff
@@ -171,61 +176,142 @@ func (idx *NetworkIndex) Overcommitted() bool {
 	return false
 }
 
-// SetNode is used to setup the available network resources. Returns
-// true if there is a collision
-func (idx *NetworkIndex) SetNode(node *Node) (collide bool, reason string) {
-
-	// COMPAT(0.11): Remove in 0.11
-	// Grab the network resources, handling both new and old
-	var networks []*NetworkResource
+// SetNode is used to initialize a node's network index with available IPs,
+// reserved ports, and other details from a node's configuration and
+// fingerprinting.
+//
+// SetNode must be idempotent as preemption causes SetNode to be called
+// multiple times on the same NetworkIndex, only clearing UsedPorts between
+// calls.
+//
+// An error is returned if the Node cannot produce a consistent NetworkIndex
+// such as if reserved_ports are unparseable.
+//
+// Any errors returned by SetNode indicate a bug! The bug may lie in client
+// code not properly validating its configuration or it may lie in improper
+// Node object handling by servers. Users should not be able to cause SetNode
+// to error. Data that cause SetNode to error should be caught upstream such as
+// a client agent refusing to start with an invalid configuration.
+func (idx *NetworkIndex) SetNode(node *Node) error {
+
+	// COMPAT(0.11): Deprecated. taskNetworks are only used for
+	// task.resources.network asks which have been deprecated since before
+	// 0.11.
+	// Grab the network resources, handling both new and old Node layouts
+	// from clients.
+	var taskNetworks []*NetworkResource
 	if node.NodeResources != nil && len(node.NodeResources.Networks) != 0 {
-		networks = node.NodeResources.Networks
+		taskNetworks = node.NodeResources.Networks
 	} else if node.Resources != nil {
-		networks = node.Resources.Networks
+		taskNetworks = node.Resources.Networks
 	}
 
-	var nodeNetworks []*NodeNetworkResource
-	if node.NodeResources != nil && len(node.NodeResources.NodeNetworks) != 0 {
-		nodeNetworks = node.NodeResources.NodeNetworks
+	// Reserved ports get merged downward. For example given an agent
+	// config:
+	//
+	// client.reserved.reserved_ports = "22"
+	// client.host_network["eth0"] = {reserved_ports = "80,443"}
+	// client.host_network["eth1"] = {reserved_ports = "1-1000"}
+	//
+	// Addresses on taskNetworks reserve port 22
+	// Addresses on eth0 reserve 22,80,443 (note 22 is also reserved!)
+	// Addresses on eth1 reserve 1-1000
+	globalResPorts := []uint{}
+
+	// COMPAT(0.11): Remove in 0.11
+	// Handle reserving ports, handling both new and old client Node
+	// layouts
```

Comment:
Is the `else if node.Reserved != nil` code path that is deprecated?

Reply:
Correct, clarifying.